java-frontend/src/main/java/org/sonar/java/reporting/InternalJavaIssueBuilder.java

```diff
@@ -0,0 +1,236 @@
+
+  @Override
+  public InternalJavaIssueBuilder withSecondaries(JavaFileScannerContext.Location... secondaries) {
+    return withSecondaries(Arrays.asList(secondaries));
+  }
+
+  @Override
+  public InternalJavaIssueBuilder withSecondaries(List<JavaFileScannerContext.Location> secondaries) {
+    requiresExistence(this.message, MESSAGE_NAME);
+    requiresUniquess(this.secondaries, "secondaries");
+    Preconditions.checkState(this.flows == null, "Cannot set flows and secondaries at the same time.");
+
+    this.secondaries = Collections.unmodifiableList(secondaries);
+    return this;
+  }
+
+  @Override
+  public InternalJavaIssueBuilder withFlows(List<List<JavaFileScannerContext.Location>> flows) {
+    requiresExistence(this.message, MESSAGE_NAME);
+    requiresUniquess(this.flows, "flows");
+    Preconditions.checkState(this.secondaries == null, "Cannot set flows and secondaries at the same time.");
+
+    this.flows = Collections.unmodifiableList(flows);
+    return this;
+  }
+
+  @Override
+  public InternalJavaIssueBuilder withCost(int cost) {
+    requiresExistence(this.message, MESSAGE_NAME);
+    requiresUniquess(this.cost, "cost");
+
+    this.cost = cost;
+    return this;
+  }
+
+  @Override
+  public void build() {
+    requiresExistence(this.rule, RULE_NAME);
+    requiresExistence(this.textSpan, TEXT_SPAN_NAME);
+    requiresExistence(this.message, MESSAGE_NAME);
+
+    if (sonarComponents == null) {
+      // can be noisy , so using trace only.
+      LOG.trace("SonarComponents is not set - discarding issue");
+      return;
+    }
+    Optional<RuleKey> ruleKey = sonarComponents.getRuleKey(rule);
+    if (!ruleKey.isPresent()) {
+      // can be noisy , so using trace only.
+      LOG.trace("Rule not enabled - discarding issue");
+      return;
+    }
+
+    NewIssue newIssue = sonarComponents.context().newIssue()
+      .forRule(ruleKey.get())
+      .gap(cost == null ? 0 : cost.doubleValue());
+
+    newIssue.at(
+      newIssue.newLocation()
+        .on(inputFile)
+        .at(inputFile.newRange(textSpan.startLine, textSpan.startCharacter, textSpan.endLine, textSpan.endCharacter))
+        .message(message));
+
+    if (secondaries != null) {
+      flows = secondaries.stream().map(Collections::singletonList).collect(Collectors.toList());
+      // Keep secondaries and flows mutually exclusive.
+      secondaries = null;
+    }
+
+    if (flows != null) {
+      for (List<JavaFileScannerContext.Location> flow : flows) {
+        newIssue.addFlow(flow.stream()
+          .map(location -> newIssue.newLocation()
+            .on(inputFile)
+            .at(range(inputFile, location))
+            .message(location.msg))
+          .collect(Collectors.toList()));
+      }
+    }
+
+    newIssue.save();
+  }
+
+  private static TextRange range(InputFile file, JavaFileScannerContext.Location location) {
+    AnalyzerMessage.TextSpan textSpan = AnalyzerMessage.textSpanFor(location.syntaxNode);
+    if (textSpan.onLine()) {
```

Comment:
I feel that this condition will never be possible. A `location` holds a `Tree` (`syntaxNode`), and it cannot be at an invalid position which would trigger an `onLine()` test to be true (`startCharacter` has to be `-1`).

What about removing it?

Reply:
We took the code from another class that was able to reach this situation. Since it is not possible currently with this new way, we can probably remove it indeed.